best-practices/sdk-principles/kotlin.md

```diff
@@ -0,0 +1,73 @@
+
+Kotlin is the primary development language for any JVM-based SDKs (Java, Scala, Groovy, Clojure, etc). Our main focus is to provide an idiomatic Kotlin API for our customers. However, upon request, we will also offer a Java package that does not use coroutines or extension functions, making it easier to consume from Java (similar to [apollo java support](https://github.com/apollographql/apollo-kotlin-java-support)).
+
+## Core SDK
+
+The Core SDK serves is a Pub/Sub channel SDK without additional plugins like VcDiff, LiveObjects, or Push Notifications.
+
+Our current pub/sub channel SDK should be split into several modules:
+
+1. `com.ably.core:api` - Public interfaces for the core SDK (includes `RestClient`, `RealtimeClient`).
+2. `com.ably.core:runtime` - Pub/sub channel SDK implementation.
+3. `com.ably.core:serialization-*` - Serialization modules (e.g., `com.ably.core:serialization-gson`).
+4. `com.ably.core:network-*` - Networking modules (e.g., `com.ably.core:network-okhttp`).
+5. `com.ably.core:plugin-*` - Plugin packages (e.g., `com.ably.core:plugin-vc-diff`).
+6. `com.ably.core:android-push` - Android Push extensions (push activation, logic). Note that the Push Admin part should be included in the public interface and implemented in `com.ably.core:runtime`.
+7. `com.ably.core:android-ext` - Android-specific extensions.
+
+### Dependency on Core SDK
+
+Other products (Spaces, Chat, Asset Tracking, etc.) should depend on `com.ably.core:api` to access the public API and potentially some internal APIs (which should be marked with a special `@InternalApi` annotation).
+
+## Dependency on Third-Party Libraries
+
+We generally try to avoid dependencies on third-party libraries, especially when it comes to networking or serialization libraries. These libraries are often opinionated and can lead to several problems: version incompatibility (e.g., OkHttp 4.x vs. OkHttp 5.x), code inconsistency, and name clashes (e.g., when a project uses `kotlinx.serialization` and we require `Gson`).
+
+When reasonable, we should create the necessary interfaces, hide dependencies behind them, and provide implementations as separate packages.
+
+## Kotlin Coroutines
+
+We want to provide an idiomatic Kotlin API for our customers, and coroutines are the idiomatic way to handle asynchronous operations and reactive programming (callbacks, subscriptions).
+
+* Async operations become Kotlin **suspending functions**.
+* Generic state changes (usually implemented with an `on()` method) are transformed into `StateFlow`.
+* Subscriptions are transformed into either `Flow` or `SharedFlow`.
+
+
+### Code Examples
+
+```kotlin
+val client = RestClient(key = "<ABLY KEY>")
+val currentStatus = client.connection.status.value
+client.connection.status.collect { nextStatus ->
+    // subsribe to status change
+    println(nextStatus)
+}
+```
+
+### Flow (Cold Flow) vs. SharedFlow (Hot Flow)
```

Comment:
more of a nitpick (feel free to ignore) - the title is Flow vs SharedFlow. The section would read nicer if it followed the same order i.e. move the "Consider using Flow when:" above "Consider using SharedFlow when:"

Reply:
+1

Reply:
thanks, fixed!